### PR TITLE
[Docs](mluOpLgamma): add docs for new operator lgamma

### PR DESCRIPTION
**Pull request by Frankd35**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

add document for new operator lgamma

## 2. Modification

add new file docs/design_docs/lgamma/lgamma.md

## 3. Test Report

not yet

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS™ Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [ ] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [ ] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [ ] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [ ] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [ ] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [*] MLU370
  - [ ] MLU590
- Job types
  - [*] BLOCK
  - [ ] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS™-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS™ Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

```bash
# The test results should contain Op name, Shape, Data type,  
#   MLU Hardware Time(us), MLU Interface Time(us), MLU IO Efficiency, 
#   MLU Compute Efficiency, and Mlu Workspace Size(Bytes)
# 
# for example:
#
# ----------- case0 -----------
# case0
# [Op name                ]: abs
# [Shape                  ]: input.shape=[1024,1024,3,4], output.shape=[1024,1024,3,4]
# [Data type]             ]: float32
# [MLU Hardware Time      ]: 15728 (us)
# [MLU Interface Time     ]: 369.008 (us)
# [MLU IO Efficiency      ]: 0.23275
# [MLU Compute Efficiency ]: 0.5
# [Mlu Workspace Size     ]: -1 (Bytes)
# 
# ----------- case1 -----------
# ...
```

Platform：MLU590
```bash
# ----------- case0 -----------
# ----------- case1 -----------
# ...
```

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: docs/design_docs/lgamma/lgamma.md [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+                                      const mluOpTensorDescriptor_t y_desc,
+                                      void *y)；
+```
+
+
+## 3 实现方案设计
+
+### 3.1 实现方案
+
+- step1 将输入张量 input 看做是一个一维数组，平均分配到所有可用的计算核上进行计算
```
Comment (mahxn0) [inline docs/design_docs/lgamma/lgamma.md:129]: 将输入张量 input 看做是一个一维数组，设其长度为input_num, 将input_num均分给每个core处理, 记作per_core_num. 
```
per_core_num = input_num / taskDim 
rem_num = input_num % taskDim
per_core_num = taskId < rem_num : per_core_num + 1 : per_core_num 
```

Comment (mahxn0) [reply docs/design_docs/lgamma/lgamma.md:129]: 可以像上面举例一样,写详细一点.

File: docs/design_docs/lgamma/lgamma.md [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+- step2 每个core执行task任务时，将该次任务对应的输入部分从 GDRAM 拷贝至 NRAM，使用兰克泽斯逼进(Lanczos approximation)法进行计算，该算法实现参考 [MindSpore 实现](https://github.com/mindspore-ai/mindspore/blob/master/mindspore/ccsrc/plugin/device/ascend/kernel/aicpu/aicpu_ops/cpu_kernel/utils/igamma_utils.h#L58)，并在本算子中实现了 simd 版本
```
Comment (mahxn0) [inline docs/design_docs/lgamma/lgamma.md:130]: 计算每个core处理的数据: balabala, 把每个core处理的数据的offset算出来, 根据nram大小计算一次最多能处理的数量deal_num, 若处理不下, 循环处理, 循环次数repeat = per_core_num / deal_num, balabala.... 
计算部分: 详细描述下兰克泽斯逼进(Lanczos approximation)法

File: docs/design_docs/lgamma/lgamma.md [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+  return std::isinf(input) ? std::numeric_limits<T>::infinity() : result;
+}
+
+```
+
+### 3.3 拆分(任务拆分，多核拆分)
+
+1、基本任务类型是Block
+
+2、对 input 进行数据拆分，当前可获得 MLU 核心数决定了任务数量 taskDim，将 input 平均分为 $\lfloor len(input) / taskDim \rfloor$ 个，每次任务分配给核心进行计算
```
Comment (mahxn0) [inline docs/design_docs/lgamma/lgamma.md:223]: 每次任务分配给核心进行计算, 此处描述不太通顺, 建议修改一下

File: docs/design_docs/lgamma/lgamma.md [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+| 表项            | 分配策略   |
+| ----------------| -----------|
+| NRAM            | 存放一次 task 中输入数据、中间数据、结果数据 |
+| WRAM            |  |
+| SRAM            |  |
+| DRAM(workspace) | 用于存放输入张量 input 的内存空间 |
+
+2、NRAM 分配策略
+
+本算子是 element-wise 算子，加速比与执行 simd 运算的元素数量成正相关，simd 一次可以计算的元素数量受到 NRAM 的限制
```
Comment (mahxn0) [inline docs/design_docs/lgamma/lgamma.md:240]: 此处是三级流水模板实现,可参考其它模板类算子修改这里

File: docs/design_docs/lgamma/lgamma.md [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+
+在算子计算过程中，有非常的多的中间变量存储，由于实现 simd ，所以每个中间变量在 NRAM 中占据的大小都是 sizeof(T) * simd 一次计算的元素数量，同时，NRAM 空间有限，NRAM 和 中间变量 buffer 的长度 buffer_size 和 buffer 数量 n_buffer 关系如下
+
+nram_size = buffer_size * n_buffer
+
+用于存储中间变量的 buffer 数与 simd 一次可以计算的元素数量成反比。想要 simd 一次能够计算更多的数据，就要对存储中间变量的 buffer 进行复用，使用尽可能少的 buffer，最终结果为至少需要 5 个 buffer 同时存储中间变量。所以，每个 task 在计算过程中除预留 5 KB 作为栈预留空间外，所有的 NRAM 资源全部平均分配给 5 个中间变量 buffer
+
```
Comment (mahxn0) [inline docs/design_docs/lgamma/lgamma.md:247]: 同上

File: docs/design_docs/lgamma/lgamma.md [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+2、流水设计
+本算子是 element-wise 算子，同时计算每一个元素的步骤较复杂，所以瓶颈主要在于计算侧，数据存取的时间在整个运算时间中占比不大；同时，软件流水需要使用 NRAM 空间，在本算子对 NRAM 需求较大，由于可以使用的 NRAM 空间有限，使用软件流水将导致一次 simd 计算中可以处理的元素数量减少，所以未采用软件流水
+
```
Comment (mahxn0) [inline docs/design_docs/lgamma/lgamma.md:250]: 模板实现已经是三级流水了

File: docs/design_docs/lgamma/lgamma.md [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+- #### 文档基本信息
+
+| 算子名称    | lgamma            |
+| ----------- | ----------------- |
+| 编制人/日期 | 代韵涛/2024-04-25 |
+
+- #### 修改记录
+
+| 版本号 | 修订人 | 修订日期   | 修订描述     |
+| ------ | ------ | ---------- | ------------ |
```
Comment (DanieeelLiu) [inline docs/design_docs/lgamma/lgamma.md:12]: 只留v1.0, 这是第一版交付

File: docs/design_docs/lgamma/lgamma.md [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+| 应用网络                                                                   | LDA模型（隐式狄利克雷分布）                                                  |
+| 输入数据类型                                                               | half, float                                                                  |
+| 输入Shape                                                                  | input: 任意Tensor                                                            |
+| 输入Layout                                                                 | Array                                                                        |
+| 输出数据类型                                                               | half, float                                                                  |
+| 输出Shape                                                                  | Array                                                                        |
+| 输出Layout                                                                 | 无要求                                                                       |
```
[the remark this answers is not in the document]
Comment (Frankd35) [reply docs/design_docs/lgamma/lgamma.md:47]: 已修改

File: docs/design_docs/lgamma/lgamma.md [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+| 需求来源                                                                   | Pytorch                                                                      |
+| 应用网络                                                                   | LDA模型（隐式狄利克雷分布）                                                  |
+| 输入数据类型                                                               | half, float                                                                  |
+| 输入Shape                                                                  | input: 任意Tensor                                                            |
+| 输入Layout                                                                 | Array                                                                        |
+| 输出数据类型                                                               | half, float                                                                  |
+| 输出Shape                                                                  | Array                                                                        |
```
[the remark this answers is not in the document]
Comment (Frankd35) [reply docs/design_docs/lgamma/lgamma.md:46]: 已修改

File: docs/design_docs/lgamma/lgamma.md [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+
+| 限制类型     | 详细说明               |
+| ------------ | ---------------------- |
+| 数据类型限制 | input仅支持half和float |
+| 布局限制     | Array                  |
+| 规模限制     | 无                     |
+| 功能限制     | 无                     |
+| 数据范围限制 | 无                     |
+| 原位限制     | 支持原位               |
+| stride限制   | 不支持stride机制       |
```
[the remark this answers is not in the document]
Comment (Frankd35) [reply docs/design_docs/lgamma/lgamma.md:89]: 已修改

File: docs/design_docs/lgamma/lgamma.md [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+```
+// spouge 算法计算方式为
+z < 0:
+  lgamma(z) = log(pi) + abs(log(sin(pi * z))) + lgamma(1 - z)
+z > 0:
+  lgamma(z+1) = (z+0.5)*log(z+a) - (z+a) + log(accm)
+
+// 为了方便 SIMD 计算，将两种情况合并计算
+
+bool need_reflect = x < 0.5;
```
[the remark this answers is not in the document]
Comment (Frankd35) [reply docs/design_docs/lgamma/lgamma.md:152]: 已修改

File: docs/design_docs/lgamma/lgamma.md [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+| 表项            | 分配策略                                     |
+| --------------- | -------------------------------------------- |
+| NRAM            | 存放一次 task 中输入数据、中间数据、结果数据 |
+| WRAM            |                                              |
+| SRAM            |                                              |
+| DRAM(workspace) | 用于存放输入张量 input 的内存空间            |
+
+2、NRAM 分配策略
+
+将 NRAM 平均分为 7 份，前四份作为 input output 的 ping-pong buffer, 后三份作为计算时使用的辅助数组
```
[the remark this answers is not in the document]
Comment (Frankd35) [reply docs/design_docs/lgamma/lgamma.md:196]: 已修改

File: docs/design_docs/lgamma/lgamma.md [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+### 3.1 实现方案
+
+lgamma 算子是 element wise 类型的算子，因此只需要按照输入数据进行数据切分，目前算子实现了多job并行机制，可按照每个core上可用nram空间对输入数据量进行拆分。
+
+- step1 对输入数据按照数据大小进行拆分，计算每个core处理的数据大小per_core_data。
+- step2 根据每个core的可用nram空间对per_core_data进行拆分，若nram放不下，则循环处理
+
+
+- step3 使用 [Spouge&#39;s approximation](https://en.wikipedia.org/wiki/Spouge%27s_approximation) 算法进行计算，考虑到精度，half 需升格为 float 完成计算再转换回 half
+
```
Comment (DanieeelLiu) [inline docs/design_docs/lgamma/lgamma.md:131]: 公式这里网页显示有点乱了，需要再看一下

File: docs/design_docs/lgamma/lgamma.md [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+  其中整数 a 的取值决定了 $c_{i}$ 的值，具体计算参考链接网页；且要求 $z > 0$
+
+  计算 Lgamma 函数则是在 Gamma 函数数值逼近算法上取 Log 后，先进行公式上的化简再计算值。同时考虑到以上 Gamma 函数的逼近算法只对输入大于 0 的情况有效，需要通过 [Euler&#39;s reflection formula](https://en.wikipedia.org/wiki/Reflection_formula) 计算 $z <= 0$ 的情况：
+
+  $ \Gamma(1-z) \Gamma(z) = \frac{\pi}{sin(\pi z)}$ => $ \Gamma(z) = \frac{\pi}{sin(\pi z) \Gamma(1-z) }$
+
+### 3.2 伪代码实现
+```
+// spouge 算法计算方式为
```
Comment (DanieeelLiu) [inline docs/design_docs/lgamma/lgamma.md:142]: 算法需要刷新下，跟目前的最新代码要对齐

File: docs/design_docs/lgamma/lgamma.md [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+见 [MLU-OPS 性能验收标准](../MLU-OPS-Performance-Acceptance-Standard.md)。
+
+## 2 算子接口设计
+
+### 2.1 参考接口
+
+- torch
+
+```python
+torch.lgamma(input, *, out=None) → Tensor
```
Comment (DanieeelLiu) [inline docs/design_docs/lgamma/lgamma.md:107]: 参考接口写 cuda kernel接口，https://github.com/pytorch/pytorch/blob/v2.1.0/aten/src/ATen/native/cuda/UnaryGammaKernels.cu#L88，参考这里

File: docs/design_docs/lgamma/lgamma.md (hunk 13 above)
Comment (Frankd35) [reply docs/design_docs/lgamma/lgamma.md:131]: 本地显示正常，暂时不知道什么原因导致的，先换成图片了

File: docs/design_docs/lgamma/lgamma.md (hunk 14 above)
Comment (Frankd35) [reply docs/design_docs/lgamma/lgamma.md:142]: 已更新

File: docs/design_docs/lgamma/lgamma.md (hunk 15 above)
Comment (Frankd35) [reply docs/design_docs/lgamma/lgamma.md:107]: 已修改

**Review by mahxn0 — APPROVED** [review]

**Review by PetrelYy — APPROVED** [review]

**Review by DanieeelLiu — APPROVED** [review]

**Pull request merged**